code_01_dataprep_bno.R

```diff
@@ -0,0 +1,204 @@
+new_zf <- collapse_zerofill(zf) # Creates a new zero-filled dataframe with a 0 marked for each checklist when the bird was not observed
+#   glimpse(zf)
+
+#### remove old data ####
+rm(data, ebd_filters, ebd_subset, zf, zf_subset)
+
+#### subset data, choose black and orange flycatcher ####
+columnsOfInterest = c("checklist_id","scientific_name","observation_count","locality","locality_id","locality_type","latitude","longitude","observation_date","time_observations_started","observer_id","sampling_event_identifier","protocol_type","duration_minutes","effort_distance_km","effort_area_ha","number_observers","species_observed","reviewed")
+
+speciesOfInterest = c("Ficedula nigrorufa")
+
+data = list(ebd, new_zf) %>% 
+  map(function(x){
+    x %>% select(one_of(columnsOfInterest)) %>% 
+      filter(scientific_name %in% speciesOfInterest)
+    # %>% dlply("scientific_name")
+    #'this above for when there are more species
+  })
+
+#### filter spatially ####
+#'load shapefiles of hill ranges
+library(sf)
+hills = st_read("hillsShapefile/Nil_Ana_Pal.shp")
+#'convert data to sf and filter spatially
+data = data %>% 
+  map(function(x){
+    x %>% st_as_sf(coords = c("longitude","latitude")) %>% 
+      st_set_crs(4326)
+  }) %>% 
+  map(function(x){
+    st_intersection(hills, x)
+  })
+
+#### presence and absence ####
+#'save a temp data file
+#save(data, file = "data.temp.rdata")
+
+load("data.temp.rdata")
+  
+#'check presence and absence in absences df, remove essentially the presences df
+data[[2]] = data[[2]] %>% filter(species_observed == F)
+
+#'in the first set, replace X, for presences, with 1
+data[[1]] = data[[1]] %>% mutate(observation_count = ifelse(observation_count == "X", "1", observation_count))
+
+#'remove records where duration is 0 and effort was > 0
+data = map(data, function(x) filter(x, duration_minutes > 0))
+
+#### custom functions ####
+
+library(sf)
+#'remove spatial geometry, preserve coordinates
+data = map(data, function(x){
+  x %>% bind_cols(data.frame(st_coordinates(.))) %>% `st_geometry<-`(NULL) %>% unclass() %>% as.data.frame()
+})
+
+#'group data by site and sampling event identifier
+#'then, sum relevant variables
+#'then, join by SEI to df of largely constant vars, such as locality
+dataGrouped = map(data, function(x){
+    x %>% group_by(sampling_event_identifier) %>% 
+      summarise_at(vars(duration_minutes, effort_distance_km, effort_area_ha), funs(sum.no.na))
+  }) %>% 
+  map2(data %>% 
+         map(function(x){select(x, sampling_event_identifier, time_observations_started, locality, locality_type, locality_id, observer_id, observation_date, scientific_name, observation_count, protocol_type, number_observers, X, Y)}), left_join)
+
+#'bind rows combining data frames, and filter
+dataGrouped = bind_rows(dataGrouped) %>% 
+  filter(duration_minutes <= 240, effort_distance_km <= 5, effort_area_ha <= 500)
+
+#'assign present or not
+dataGrouped = mutate(dataGrouped, pres_abs = observation_count >= 1)
+  
```

Comment:
I think it would be good to add the sub-sampling bit here? Don't you think?

Reply:
yes